{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import alignparse.targets\n",
    "import alignparse.minimap2\n",
    "import alignparse.consensus\n",
    "from alignparse.constants import CBPALETTE\n",
    "\n",
    "import dms_variants.utils\n",
    "import dms_variants.codonvarianttable\n",
    "\n",
    "from plotnine import *\n",
    "from io import StringIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Set directory\n",
    "outdir = 'ap_out' # Put './' if it should be stored at the same place as this notebook.\n",
    "\n",
    "#Input files\n",
    "target_file = 'ap_in/file.gb'\n",
    "parse_specs_file = 'file.yaml'\n",
    "ccs_folder = './ap_in' # Folder where ccs file(s) are found\n",
    "\n",
    "#Variables\n",
    "take_reverse_complement = False\n",
    "error_rate_floor = 1e-7  # error rates < this set to this\n",
    "error_cutoff = 1e-4\n",
    "gen_sequence = 'ATGCATTCTCAAAA...' # From snapgene\n",
    "\n",
    "#Output file\n",
    "create_lookup_table = False\n",
    "lookup_table_filename = 'out.csv'\n",
    "num_of_indels = 'number_of_indels == 0' # in the final lookup tables\n",
    "\n",
    "create_variants_lookup_table = False\n",
    "variants_lookup_table_filename = 'var_out.csv'\n",
    "\n",
    "#Visualization\n",
    "display_lookup_table = False\n",
    "display_variants_lookup_table = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(outdir, exist_ok=True)\n",
    "current_path = os.getcwd()\n",
    "outdir = os.path.join(current_path,outdir)\n",
    "\n",
    "targets = alignparse.targets.Targets(\n",
    "    seqsfile=target_file, #name of the .gb file\n",
    "    feature_parse_specs=parse_specs_file #name of the .yaml file\n",
    ")\n",
    "\n",
    "fastqs = [os.path.join(ccs_folder,d) for d in os.listdir(ccs_folder) if d.endswith('fastq') or d.endswith('fastq.gz')]\n",
    "run_names = [os.path.basename(os.path.splitext(name)[0]) for name in fastqs]\n",
    "libraries = run_names.copy()\n",
    "\n",
    "pacbio_runs = pd.DataFrame( # table with name library and fastq file found \n",
    "    {'name': run_names,\n",
    "    'library': libraries,\n",
    "    'fastq': fastqs\n",
    "    })\n",
    "\n",
    "mapper = alignparse.minimap2.Mapper(alignparse.minimap2.OPTIONS_CODON_DMS)\n",
    "\n",
    "readstats, aligned, filtered = targets.align_and_parse(\n",
    "    df=pacbio_runs,\n",
    "    mapper=mapper,\n",
    "    outdir=outdir,\n",
    "    name_col='name',\n",
    "    queryfile_col='fastq',\n",
    "    group_cols=['library'],\n",
    "    overwrite=True,  # overwrite any existing output\n",
    ")\n",
    "\n",
    "if take_reverse_complement:\n",
    "    aligned_df = (aligned[targets.target_names[0]]\n",
    "                  .assign(barcode=lambda x: x['barcode_sequence'].map(dms_variants.utils.reverse_complement),\n",
    "                          barcode_error=lambda x: np.clip(1 - x['barcode_accuracy'],error_rate_floor, None),\n",
    "                          gene_error=lambda x: np.clip(1 - x['gene_accuracy'],error_rate_floor, None),\n",
    "                          retained=lambda x: ((x['gene_error'] <= error_cutoff) & \n",
    "                                              (x['barcode_error'] <= error_cutoff)),\n",
    "                         )\n",
    "                 )\n",
    "else:\n",
    "    aligned_df = (aligned[targets.target_names[0]]\n",
    "              .assign(barcode=lambda x: x['barcode_sequence'],\n",
    "                      barcode_error=lambda x: np.clip(1 - x['barcode_accuracy'],error_rate_floor, None),\n",
    "                      gene_error=lambda x: np.clip(1 - x['gene_accuracy'],error_rate_floor, None),\n",
    "                      retained=lambda x: ((x['gene_error'] <= error_cutoff) & \n",
    "                                          (x['barcode_error'] <= error_cutoff)),\n",
    "                     )\n",
    "             )\n",
    "\n",
    "aligned_df = alignparse.consensus.add_mut_info_cols(\n",
    "    aligned_df,\n",
    "    mutation_col='gene_mutations',\n",
    "    n_indel_col='n_indels'\n",
    ").assign(has_indel=lambda x: x['n_indels'] > 0)\n",
    "\n",
    "#aligned_df = aligned_df[['library', 'name', 'query_name', 'barcode', 'gene_mutations','barcode_accuracy', 'gene_accuracy', 'barcode_error', 'gene_error', 'retained', 'n_indels', 'has_indel']]\n",
    "\n",
    "consensus, dropped = alignparse.consensus.simple_mutconsensus(\n",
    "    aligned_df.query('retained'),\n",
    "    group_cols=('library', 'barcode'),\n",
    "    mutation_col='gene_mutations',\n",
    ")\n",
    "\n",
    "consensus = alignparse.consensus.add_mut_info_cols(\n",
    "    consensus,\n",
    "    mutation_col='gene_mutations',\n",
    "    sub_str_col='substitutions',\n",
    "    n_indel_col='number_of_indels',\n",
    "    overwrite_cols=True\n",
    ")\n",
    "\n",
    "lookup_table = consensus.query(num_of_indels)\n",
    "\n",
    "if create_lookup_table:\n",
    "    lookup_table.to_csv(os.path.join(outdir,lookup_table_filename))\n",
    "\n",
    "if display_lookup_table:\n",
    "    display(lookup_table)    \n",
    "\n",
    "variants = dms_variants.codonvarianttable.CodonVariantTable(\n",
    "    barcode_variant_file = StringIO(lookup_table.to_csv()),\n",
    "    geneseq = gen_sequence,\n",
    ")\n",
    "\n",
    "variants_lookup_table = variants.barcode_variant_df\n",
    "\n",
    "if create_variants_lookup_table:\n",
    "    variants_lookup_table.to_csv(os.path.join(outdir,variants_lookup_table_filename))\n",
    "    \n",
    "if display_variants_lookup_table:\n",
    "    display(variants_lookup_table)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
